org.eclipse.jdt.core.javac/src/org/eclipse/jdt/internal/javac/UnusedTreeScanner.java

```diff
@@ -162,7 +162,13 @@ private boolean isPrivateDeclaration(Tree tree) {
 		if (tree instanceof JCClassDecl classTree) {
 			return (classTree.getModifiers().flags & Flags.PRIVATE) != 0;
 		} else if (tree instanceof JCMethodDecl methodTree) {
-			return !isConstructor(methodTree) && (methodTree.getModifiers().flags & Flags.PRIVATE) != 0;
+			if (isConstructor(methodTree)) {
+				if (hasPackageVisibleConstructor(methodTree.sym.owner)) {
+					return (methodTree.getModifiers().flags & Flags.PRIVATE) != 0;
```

Comment:
I would have checked if constructor && has private flag first, then proceed to detect non private constructors. Seems like less CPU cycles involved in most cases

Reply:
fixed by https://github.com/eclipse-jdtls/eclipse-jdt-core-incubator/commit/5cadbcc0c5a3cfcf628141f29e686873990b52d2